### PR TITLE
Streamline KD & QAD transformers Trainers

### PR DESCRIPTION
**Pull request by AAnoosheh**
## What does this PR do?

**Type of change:** ? Refactor and stabilization

**Overview:** 
* Enforce use of FSDP-2 on KD and QAD trainers in HF plugins/examples so that we can remove multiple restrictions

## Usage

```python
# Add a code snippet demonstrating how to use this
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No

### CODE REVIEW

File: examples/llm_qat/README.md [hunk 1]
```diff
@@ -245,7 +237,7 @@ You could also add your own customized quantization format to `CUSTOM_QUANT_CFG`
 
 > **_NOTE:_** `launch.sh` defaults to use `LlamaDecoderLayer` as the transformer layer class. If your model uses a different class, you need to pass `--fsdp_transformer_layer_cls_to_wrap <your_layer_class>` to the `launch.sh` script. For example, for `Qwen/Qwen3-8B`, specify `--fsdp_transformer_layer_cls_to_wrap Qwen3DecoderLayer` as an additional argument.
 
-> **_NOTE:_** The script defaults to using FSDP1. To use FSDP2, pass "--use_fsdp2 True" to the `launch.sh` script. Note that FSDP2 is less stable than FSDP1 currently. Use it with caution.
+> **_NOTE:_** The script defaults to using FSDP1. To use FSDP2, pass "--backend=fsdp2" to the `launch.sh` script. Note that FSDP2 is less stable than FSDP1 currently. Use it with caution.
```
Comment (kevalmorabia97) [inline examples/llm_qat/README.md]: Is this statement still valid? `Note that FSDP2 is less stable than FSDP1 currently. Use it with caution.`

Comment (AAnoosheh) [reply examples/llm_qat/README.md]: I doubt it, but I don't have proof. I don't have proof that it is less stable either.

File: modelopt/torch/distill/plugins/huggingface.py [hunk 2]
```diff
@@ -49,44 +53,40 @@ def save_model(
...
-                    save_function=self.accelerator.save,
-                    state_dict=state_dict,
-                )
-                self.processing_class.save_pretrained(output_dir)
-                torch.save(modelopt_state, f"{output_dir}/modelopt_state.pth")
-        else:
-            model = model.export() if export_student else model
-            super().save_model(output_dir, _internal_call, *args, **kwargs)
+        with model.hide_teacher_model(), model.hide_loss_modules(enable=not _internal_call):
+            if _internal_call:
```
Comment (realAsma) [inline modelopt/torch/distill/plugins/huggingface.py:83]: Why do we need to save teacher if `_internal_call` = True? Can we avoid saving teacher irrespective  of `_internal_call` value? This will accelerate checkpoint save/load (you would likely need to hide teach during the final checkpoint load as well)

File: examples/llm_qat/README.md [hunk 3]
```diff
@@ -276,10 +268,11 @@ To perform QAD with logits loss, run:
    --quant_cfg NVFP4_DEFAULT_CFG \
    --do_train True \
    --output_dir llama-qad \
-   --distill True
+   --distill True \
+   --backend fsdp2
 ```
 
-> **_NOTE:_** QAD currently requires quantization to be applied before the FSDP wrapper. Training is not supported for models that exceed single GPU memory capacity.
+> **_NOTE:_** QAD doesn't support original FSDP - only FSDP2. It also requires quantization to be applied before the FSDP wrapper.
```
Comment (realAsma) [inline examples/llm_qat/README.md]:
```suggestion
> **_NOTE:_** QAD doesn't support FSDP1 (https://docs.pytorch.org/docs/stable/fsdp.html)  - only FSDP2. It also requires quantization to be applied before the FSDP wrapper.
```

File: modelopt/torch/distill/plugins/huggingface.py (hunk 2 above)
Comment (AAnoosheh) [reply modelopt/torch/distill/plugins/huggingface.py:83]: I think you misread it. Teacher is indeed always hidden.

File: examples/llm_qat/main.py
Comment (realAsma) [inline examples/llm_qat/main.py:1]: nice! looks clean

File: examples/llm_qat/README.md [hunk 4]
```diff
@@ -276,10 +268,11 @@ To perform QAD with logits loss, run:
    --quant_cfg NVFP4_DEFAULT_CFG \
    --do_train True \
    --output_dir llama-qad \
-   --distill True
+   --distill True \
+   --backend fsdp2
 ```
 
-> **_NOTE:_** QAD currently requires quantization to be applied before the FSDP wrapper. Training is not supported for models that exceed single GPU memory capacity.
+> **_NOTE:_** QAD doesn't support FSDP1 (<https://docs.pytorch.org/docs/stable/fsdp.html>)  - only FSDP2. It also requires quantization to be applied before the FSDP wrapper.
```
Comment (realAsma) [inline examples/llm_qat/README.md]: Can we remove `It also requires quantization to be applied before the FSDP wrapper.` -> This is a detail and could confuse users.

```suggestion
> **_NOTE:_** QAD doesn't support FSDP1 (<https://docs.pytorch.org/docs/stable/fsdp.html>)  - only FSDP2.
```

File: examples/llm_qat/README.md
Comment (realAsma) [inline examples/llm_qat/README.md:1]: Can we make `fsdp2` the default in the scripts and avoid having to specify `--backend fsdp2`?

File: modelopt/torch/quantization/plugins/transformers_trainer.py [hunk 5]
```diff
@@ -414,23 +409,19 @@ def save_model(
...
-            if export_student:
-                model = self.accelerator.unwrap_model(self.model)
-                model = model.export()
-            return QATTrainer.save_model(self, output_dir, _internal_call, *args, **kwargs)
-        return KDTrainer.save_model(
-            self, output_dir, _internal_call, export_student, *args, **kwargs
-        )
+            model = self.accelerator.unwrap_model(self.model)
+            with model.hide_teacher_model(), model.hide_loss_modules(enable=not _internal_call):
+                return QATTrainer.save_model(self, output_dir, _internal_call, *args, **kwargs)
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformers_trainer.py]: @AAnoosheh when does `mtd.export` happens now?

Comment (AAnoosheh) [reply modelopt/torch/quantization/plugins/transformers_trainer.py]: It doesn't actually need to. Hiding the teacher is enough for saving state.

Since export is a permanent in-place operation, I'd rather avoid it if not absolutely necessary.

Comment (realAsma) [reply modelopt/torch/quantization/plugins/transformers_trainer.py]: @AAnoosheh okay what happens when we try to restore the final model -> will this be restored with distillation wrapping?

Comment (AAnoosheh) [reply modelopt/torch/quantization/plugins/transformers_trainer.py]: KD mode no longer saves itself in modelopt state, so it will be identical to exported one

File: modelopt/torch/quantization/plugins/transformers_trainer.py [hunk 6]
```diff
@@ -414,23 +409,19 @@ def save_model(
...
+        else:
+            return KDTrainer.save_model(self, output_dir, _internal_call, *args, **kwargs)
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformers_trainer.py]: Do we need `def save_model(` for `QADTrainer`?
MRO for QADTrainer.save_model is `QATTrainer.save_model -> KDTrainer.save_model -> HF Trainer.svae_model right? What in particular is being achieved here now?

Comment (realAsma) [reply modelopt/torch/quantization/plugins/transformers_trainer.py]: If we don't respect this MRO, the fix in https://github.com/NVIDIA/Model-Optimizer/pull/546/files wont be applied for QAD checkpoints

Comment (realAsma) [reply modelopt/torch/quantization/plugins/transformers_trainer.py]: If we  remove `save_model`  overwriting from here - does this still work? If so can we remove it?

Comment (AAnoosheh) [reply modelopt/torch/quantization/plugins/transformers_trainer.py]: Yep it works. Done

File: modelopt/torch/quantization/plugins/transformers_trainer.py [hunk 7]
```diff
@@ -385,52 +382,25 @@ def __init__(
...
-            self.model.add_adapter(self.args.lora_config)
-            print_rank_0("Lora adapter added.")
         self._convert_to_distillation_model()
+        model = self.accelerator.unwrap_model(self.model)
+        with model.hide_teacher_model(), model.only_student_forward():
+            if self.quant_cfg is not None and not is_quantized(self.model):
+                self._quantize_model()
+            if getattr(self.args, "lora_config", None) is not None:
+                self.model.add_adapter(self.args.lora_config)
+                print_rank_0("Lora adapter added.")
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformers_trainer.py]: Can we do this?

The current implementation is still in-efficient - We should quantize the model only after wrapping it in FSDP2.

```suggestion

       def _quantize(self):
           model = self.accelerator.unwrap_model(self.model)
           with model.hide_teacher_model(), model.only_student_forward():
                return super()._quantize()
```

File: modelopt/torch/quantization/plugins/transformers_trainer.py [hunk 8]
```diff
@@ -385,52 +382,25 @@ def __init__(
...
         assert distill_config is not None, "`distill_config` is required for QAD."
         self.distill_config = distill_config
-
         super().__init__(*args, **kwargs)
 
-        # Note: QAD doesn't work with FSDP wrapped model. We quantize model before the wrapper.
-        # The drawback is that we can't train a model that is bigger than a single GPU memory.
-        # And memory efficient loading doesn't work.
+        # Note: FSDP memory efficient loading doesn't work.
         self.model.cuda()
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformers_trainer.py]: This should not be needed -> model will be moved to the correct device during FSDP wrapping by accelerate.

```suggestion
```

File: modelopt/torch/quantization/plugins/transformers_trainer.py [hunk 9]
```diff
@@ -385,52 +382,25 @@ def __init__(
...
-        if self.quant_cfg is not None and not is_quantized(self.model):
-            self._quantize_model()
-        if getattr(self.args, "lora_config", None) is not None:
-            self.model.add_adapter(self.args.lora_config)
-            print_rank_0("Lora adapter added.")
         self._convert_to_distillation_model()
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformers_trainer.py]: Can we move this to  KDTrainer `__init__`?

```suggestion

```

File: modelopt/torch/quantization/plugins/transformers_trainer.py [hunk 10]
```diff
@@ -385,52 +382,25 @@ def __init__(
         """Initialize the trainer with modelopt states."""
         assert distill_config is not None, "`distill_config` is required for QAD."
         self.distill_config = distill_config
-
         super().__init__(*args, **kwargs)
 
-        # Note: QAD doesn't work with FSDP wrapped model. We quantize model before the wrapper.
-        # The drawback is that we can't train a model that is bigger than a single GPU memory.
-        # And memory efficient loading doesn't work.
+        # Note: FSDP memory efficient loading doesn't work.
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformers_trainer.py]: wont be needed if we do https://github.com/NVIDIA/Model-Optimizer/pull/708/files#r2677776721

```suggestion
```

File: modelopt/torch/quantization/plugins/transformers_trainer.py [hunk 11]
```diff
@@ -385,52 +382,25 @@ def __init__(
...
+            if getattr(self.args, "lora_config", None) is not None:
+                self.model.add_adapter(self.args.lora_config)
+                print_rank_0("Lora adapter added.")
 
     def _convert_to_distillation_model(self):
         """Convert the model to a distillation model."""
-        # We don't need any save/restore feature of the distallation mode, so we skip it here.
-        _convert_for_kd(self.model, KDLossConfig(**self.distill_config))
+        mtd.convert(self.model, mode=[("kd_loss", self.distill_config)])
         print_rank_0("Distillation model created.")
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformers_trainer.py]: This does not seem to be specific to QAD. Should we move this to `KDTrainer` and call this during `KDTrainer` `__init__`?

**Review by realAsma — APPROVED** [review]
Looks great!!
